autopilot/contractor.go

```diff
@@ -386,68 +402,63 @@ func (c *contractor) runContractFormations(cfg Config, budget *types.Currency, r
 			continue
 		}
 
-		// fetch full contract from db.
-		formedContract, err := c.ap.bus.Contract(contract.ID())
-		if err != nil {
-			return nil, err
-		}
-
 		// add contract to contract set
-		formed = append(formed, formedContract)
+		formed = append(formed, contract)
 
 		missing--
 	}
 
 	return formed, nil
 }
 
-func (c *contractor) renewContract(cfg Config, toRenew bus.Contract, renterKey consensus.PrivateKey, renterAddress types.UnlockHash, renterFunds, hostCollateral types.Currency) (rhpv2.ContractRevision, error) {
+func (c *contractor) renewContract(cfg Config, toRenew rhpv2.ContractRevision, hostIP string, renterKey consensus.PrivateKey, renterAddress types.UnlockHash, renterFunds, hostCollateral types.Currency) (rhpv2.ContractRevision, error) {
 	// handle contract locking
-	revision, err := c.ap.bus.AcquireContract(toRenew.ID(), contractLockingDurationRenew)
+	locked, err := c.ap.bus.AcquireContract(toRenew.ID(), contractLockingDurationRenew)
 	if err != nil {
 		return rhpv2.ContractRevision{}, nil
 	}
+	if !locked {
+		return rhpv2.ContractRevision{}, errors.New("contract is currently locked")
+	}
 	defer c.ap.bus.ReleaseContract(toRenew.ID())
 
 	// fetch host settings
-	scan, err := c.ap.worker.RHPScan(toRenew.HostKey(), toRenew.HostIP, 0)
+	scan, err := c.ap.worker.RHPScan(toRenew.HostKey(), hostIP, 0)
 	if err != nil {
 		c.logger.Debugw(
 			fmt.Sprintf("failed scan, err: %v", err),
-			"hk", toRenew.HostKey(),
+			"hk", toRenew.HostKey,
 		)
 		return rhpv2.ContractRevision{}, nil
 	}
 
 	// prepare the renewal
-	endHeight := c.currentPeriod + cfg.Contracts.Period + cfg.Contracts.RenewWindow
-	fc, cost, finalPayment, err := c.ap.worker.RHPPrepareRenew(revision, renterKey, toRenew.HostKey(), renterFunds, renterAddress, endHeight, scan.Settings)
-	if err != nil {
-		return rhpv2.ContractRevision{}, nil
-	}
-
-	// fund the transaction
-	txn := types.Transaction{FileContracts: []types.FileContract{fc}}
-	toSign, parents, err := c.ap.bus.WalletFund(&txn, cost)
-	if err != nil {
-		_ = c.ap.bus.WalletDiscard(txn) // ignore error
-		return rhpv2.ContractRevision{}, err
-	}
+	signingFn := func(fc types.FileContract, cost types.Currency) ([]types.Transaction, func() error, error) {
```

Comment:
maybe this is a good place to discuss how the contractor does contract formations and renewals, in #42 I have it go through the bus first and then push the (signed) txn to the workers